crates/evm/src/rpc_helpers/filter.rs

```diff
@@ -555,6 +555,13 @@ impl<'de> Deserialize<'de> for Filter {
     }
 }
 
+impl std::hash::Hash for Filter {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
```

Comment:
I guess this is used for constructing a hashmap over Filter. are we sure we have to impl. this trait? how did reth handle log subscriptions for example?

Reply:
Check the following code they are using:
```
/// Returns a stream that yields all logs that match the given filter.
    fn log_stream(&self, filter: FilteredParams) -> impl Stream<Item = Log> {
        BroadcastStream::new(self.chain_events.subscribe_to_canonical_state())
            .map(move |canon_state| {
                canon_state.expect("new block subscription never ends").block_receipts()
            })
            .flat_map(futures::stream::iter)
            .flat_map(move |(block_receipts, removed)| {
                let all_logs = logs_utils::matching_block_logs_with_tx_hashes(
                    &filter,
                    block_receipts.block,
                    block_receipts.tx_receipts.iter().map(|(tx, receipt)| (*tx, receipt)),
                    removed,
                );
                futures::stream::iter(all_logs)
            })
    }
```
As far as I understood, they have some sort of internal subscription mechanism to changes to the chain such as new block or reorgs. They first subscribe to this, and for each subscription, they are querying all receipts, and applying the filter on top of them. So the only difference (other than implementation details) is that we store subscriptions by filter, which allows us to group the same filters into the same operation. The reason I went with this is, most of the time subscriptions will be for many different types of filters. But high percentage of the subscriptions will be the same for some popular DEX like Uniswap. So I though we can get away with applying the same filter once.

Being said that, I also understand that when assumptions are forced on the behavior implemented, it doesn't yield the best result. Hence I am open to changing this to a simple vector of subscriptions and iterating over it.

Reply:
One thing we can do is, once the soft confirmation signal arrives, SubscriptionManager can query all the logs in the block and send them all to each subscriber. And then subscribers can apply their filters onto these logs themselves. This way we don't have to have hash map, nor Hash trait on the filter.